### ISSUE TITLE
Memory leak in Openlayers 6 beta

### ISSUE THREAD

**Issue by panuchka** [openlayers/openlayers#9556]
**Versions affected**
At least versions 6.0.0-beta.8 and 6.0.0-beta.7.
This problem is not present in 5.3.0.

**Description**
I have a situation where I need to quickly create and destroy maps. After upgrading to Openlayers 6 beta, I encountered a page crash after a while. The problem is that canvas/map is not destroyed, something is keeping it alive. 

Map has three layers:
- Two TileLayers
- One VectorLayer (using the vector mode)

If I don't initialize the vector layer, the map gets destroyed just fine (so the tile layers are ok). If I initialize the vector layer with just the default styles, map stays in memory after it is supposed to be destroyed. 

**To Reproduce**
Map is embedded in Vue.js component. When Vue.js component gets destroyed, the map instance is destroyed before Vue scraps the dom. I clear the vector layer with `layer.getSource().clear()`, `layer.setSource(undefined)` and `map.removeLayer(layer)`. Then I try to destroy map with `map.setTarget(null)` and `map = null`. However, _something_ stays in the memory. I've been hunting the cause with browser memory snapshot, and it clearly shows that canvas is not destroyed:

![image](https://user-images.githubusercontent.com/6607517/57851662-d23fe980-77e9-11e9-907f-c7613cb1840c.png)

I'm kind of suspicious of the LabelCache doing something dirty, since it was added to the project this year. If it doesn't ring any bells, I can try to create a live example of this problem. 

**Expected behavior**
Map/canvas does not stay in memory when it's destroyed.

**Comment by fredj** [comment]
It would be nice if you can create a minimal example.

**Comment by panuchka** [comment]
Yeah I can do that. Do you happen to have any templates for OL6 for code sandbox?

**Comment by fredj** [comment]
The easiest is tho start with [an example](https://openlayers.org/en/master/examples/simple.html) and click the "Edit" link above the map (on the right). This will create a codesandbox.io

**Comment by ahocevar** [comment]
@panuchka Vector layers register a listener on the label cache. Maybe calling `dispose()` on vector layers helps? Also, have you triggered a garbage collection before taking the memory snapshot? When the vector layer is no longer referenced by anything, that registered listener should not matter.

**Comment by panuchka** [comment]
@ahocevar Tried everything you suggested but not avail. 

Here is a _somewhat_ replication of the problem I have with my current project in code sandbox. The example destroys the old map instance and creates a new one with one tile and vector layer every second.
Editable: https://codesandbox.io/s/simple-grm6o
Full screen (use this to snapshot/record performance): https://grm6o.codesandbox.io/

I hope I haven't done something stupid :P But this is how it looks on my chrome's perf monitor:
![image](https://user-images.githubusercontent.com/6607517/57856665-2ea90600-77f6-11e9-91af-1bbbed276576.png)

I also tested this with changing to OL 5.3 in package json and the graph looks like this:
![image](https://user-images.githubusercontent.com/6607517/57857294-7aa87a80-77f7-11e9-8f7f-671f74bdf29e.png)

Ping @fredj

**Comment by ahocevar** [comment]
Ok, I see what's going on: the map renderer has no chance to schedule the removal of layer renderers, which would unregister the listener for label cache changes. See https://codesandbox.io/s/simple-zlehr, which disposes the layer renderer, which makes the leak go away.

I'll see if we can fix that in the library.

**Comment by ahocevar** [comment]
The issue turned out to be something completely different. When #9561 is merged, it will be enough again to call `map.setTarget(null)` to release all resources.

**Comment by panuchka** [comment]
Thank you for the fix! 👍 Any estimate on beta 9 release?

**Comment by panuchka** [comment]
Hi again!

Tried the example we tested/created with beta 9 (https://codesandbox.io/s/simple-114uh) and noticed that it might still leak stuff. Not sure if this was always the case with beta 8 too but I missed it? The problem seems to be that canvas is still not destroyed properly. The leak is very subtle and memory is not increasing that fast. I let the example run for a minute and got this:

![image](https://user-images.githubusercontent.com/6607517/58692077-87e86c00-8396-11e9-865b-4c092a902794.png)

As you can see, it's still going up, even though almost everything is freed. And the memory snapshot says that labelCache is still there 🤔 

![image](https://user-images.githubusercontent.com/6607517/58693067-9cc5ff00-8398-11e9-9309-28fa0eb9a5e5.png)

Could a short re-investigation be done? @ahocevar 
5.3.0. seems to work fine on this regard.

**Comment by ahocevar** [comment]
I cannot reproduce your results with the codesandbox you provided. I took four heap snapshots within 5 minutes of running your sandbox, and this is what I get:
![Untitled](https://user-images.githubusercontent.com/211514/58705299-09e48f00-83af-11e9-9624-593e10754e11.png)

**Comment by panuchka** [comment]
Sorry, the codesandbox I provided had 5.3.0 in package json (I probably copy pasted wrong url 😢 ).
Can you reproduce it with beta 9, here is a proper link to full screen version: https://kmgxf.codesandbox.io

**Comment by ahocevar** [comment]
Thanks for testing and reporting this @panuchka. There was indeed a remaining issue, which I just fixed in #9637.

**Comment by sweet369Happpy** [comment]
@panuchka @ahocevar hello, may i know any progress about this issues（Memory leak in Openlayers 6 beta https://github.com/openlayers/openlayers/issues/9556), we hit the same issues with Memory leak... could you share some suggestion for us, many thanks!!!
